// vtex/lib/lib/constants/InstallStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallStatus = void 0;
var InstallStatus;
(function (InstallStatus) {
    InstallStatus["ACCEPTED_TERMS"] = "installed_by_accepted_terms";
    InstallStatus["ALREADY_PURCHASED"] = "already_purchased";
    InstallStatus["CHECK_TERMS"] = "check_terms";
    InstallStatus["FREE"] = "installed_free";
    InstallStatus["OWN_APP"] = "own_app";
    InstallStatus["OWN_REGISTRY"] = "installed_from_own_registry";
    InstallStatus["PREVIOUS_PURCHASE"] = "installed_by_previous_purchase";
    InstallStatus["PUBLIC_REGISTRY"] = "public_app";
    InstallStatus["USER_HAS_NO_BUY_APP_LICENSE"] = "no_buy_app_license";
    InstallStatus["USER_HAS_NO_INSTALL_APP_LICENSE"] = "no_install_app_license";
    InstallStatus["AREA_UNAVAILABLE"] = "area_unavailable";
    InstallStatus["CONTRACT_NOT_FOUND"] = "app_store_contract_not_found";
    InstallStatus["ACCOUNT_NOT_SPONSORED"] = "account_not_sponsored_by_vendor";
})(InstallStatus = exports.InstallStatus || (exports.InstallStatus = {}));
