// vtex/lib/commands/release.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const CustomCommand_1 = require("../api/oclif/CustomCommand");
const release_1 = tslib_1.__importStar(require("../modules/release"));
const Colors_1 = require("../api/constants/Colors");
class Release extends CustomCommand_1.CustomCommand {
    async run() {
        const { args: { releaseType, tagName }, } = this.parse(Release);
        await release_1.default(releaseType, tagName);
    }
}
exports.default = Release;
Release.description = '(Only for git users) Bumps the app version, commits, and pushes to remote the app in the current directory.';
Release.examples = [
    `${Colors_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex release')}`,
    `${Colors_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex release')} patch`,
    `${Colors_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex release')} patch beta`,
    `${Colors_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex release')} minor stable`,
    `${Colors_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex release')} pre`,
];
Release.flags = {
    ...CustomCommand_1.CustomCommand.globalFlags,
};
Release.args = [
    {
        name: 'releaseType',
        required: false,
        default: 'patch',
        options: [...Object.keys(release_1.releaseTypeAliases), ...release_1.supportedReleaseTypes],
        description: `Release type.`,
    },
    { name: 'tagName', required: false, default: 'beta', options: release_1.supportedTagNames, description: `Tag name.` },
];
