// vtex/lib/modules/setup/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONTENT_ESLINT_RC_BUILDERS = exports.CONTENT_BASE_ESLINT_RC = exports.CONTENT_PRETTIER_RC = exports.CONTENT_ESLINT_IGNORE = exports.DEPENDENCIES = exports.BUILDERS_WITH_TOOLING = exports.BUILDERS_WITH_TYPES = void 0;
// Builders that use typing information
exports.BUILDERS_WITH_TYPES = ['react', 'node'];
// Builders that demand tooling installation
exports.BUILDERS_WITH_TOOLING = ['react', 'node'];
// Map of dependencies for the whole project and for each builder
exports.DEPENDENCIES = {
    // Common dependencies between projects
    common: {
        '@vtex/prettier-config': '^0.1.4',
        eslint: '^7.12.1',
        'eslint-config-vtex': '^12.3.2',
        husky: '^4.2.3',
        'lint-staged': '^10.1.1',
        prettier: '^2.0.2',
        typescript: '^3.8.3',
    },
    react: {
        'eslint-config-vtex-react': '^6.3.2',
    },
    node: {
        '@types/node': '^12.12.21',
    },
};
exports.CONTENT_ESLINT_IGNORE = `
node_modules/
coverage/
*.snap.ts
`;
exports.CONTENT_PRETTIER_RC = `@vtex/prettier-config`;
exports.CONTENT_BASE_ESLINT_RC = {
    extends: 'vtex',
    root: true,
    env: {
        node: true,
    },
};
exports.CONTENT_ESLINT_RC_BUILDERS = {
    react: {
        extends: 'vtex-react/io',
    },
};
