// @vtex/cli-plugin-deps/build/commands/deps/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const vtex_1 = require("vtex");
const update_1 = tslib_1.__importDefault(require("../../modules/deps/update"));
class DepsUpdate extends vtex_1.CustomCommand {
    async run() {
        const { raw } = this.parse(DepsUpdate);
        const allArgs = this.getAllArgs(raw);
        await update_1.default(allArgs);
    }
}
exports.default = DepsUpdate;
DepsUpdate.description = `Updates a dependency of the current workspace. If not specified which dependency, it updates all of them.`;
DepsUpdate.examples = [
    `${vtex_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex deps update')}`,
    `${vtex_1.ColorifyConstants.COMMAND_OR_VTEX_REF('vtex deps update')} vtex.service-example@0.0.1`,
];
DepsUpdate.flags = {
    ...vtex_1.CustomCommand.globalFlags,
};
DepsUpdate.strict = false;
DepsUpdate.args = [
    {
        name: 'appId',
        required: false,
        description: `Name and version of the app ${vtex_1.ColorifyConstants.ID('({vendor}.{appname}@{x.x.x})')} to update.`,
    },
    {
        name: 'ithAppId',
        required: false,
        multiple: true,
        description: `Names and versions of the multiple apps ${vtex_1.ColorifyConstants.ID('({vendor}.{appname}@{x.x.x})')} to update.`,
    },
];
