// vtex/lib/api/conf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCluster = exports.saveCluster = exports.Region = exports.getEnvironment = exports.saveNumberOfReactLinks = exports.getNumberOfReactLinks = exports.saveLastLinkReactDate = exports.getLastLinkReactDate = exports.saveNextFeedbackDate = exports.getNextFeedbackDate = exports.hasStickyHost = exports.getStickyHost = exports.saveStickyHost = exports.saveEnvironment = exports.ENV_DEFAULT_VALUE = exports.CLUSTER_DEFAULT_VALUE = exports.Environment = exports.configDir = void 0;
const tslib_1 = require("tslib");
const configstore_1 = tslib_1.__importDefault(require("configstore"));
const path_1 = require("path");
const package_json_1 = require("../../package.json");
const conf = new configstore_1.default(package_json_1.name);
exports.configDir = path_1.dirname(conf.path);
var Environment;
(function (Environment) {
    Environment["Production"] = "prod";
})(Environment = exports.Environment || (exports.Environment = {}));
exports.CLUSTER_DEFAULT_VALUE = '';
exports.ENV_DEFAULT_VALUE = Environment.Production;
exports.saveEnvironment = (env) => conf.set('env', env);
exports.saveStickyHost = (appName, stickyHost) => conf.set(`apps.${appName}.sticky-host`, { stickyHost, lastUpdated: Date.now() });
exports.getStickyHost = (appName) => conf.get(`apps.${appName}.sticky-host`);
exports.hasStickyHost = (appName) => conf.has(`apps.${appName}.sticky-host`);
exports.getNextFeedbackDate = () => conf.get('_nextFeedbackDate');
exports.saveNextFeedbackDate = (date) => conf.set('_nextFeedbackDate', date);
const REACT_LINK_MESSAGE_DISPLAY = '_lastLinkReactDate';
exports.getLastLinkReactDate = () => conf.get(REACT_LINK_MESSAGE_DISPLAY);
exports.saveLastLinkReactDate = (date) => conf.set(REACT_LINK_MESSAGE_DISPLAY, date);
const REACT_LINKS_COUNTER = '_numberOfReactLinks';
exports.getNumberOfReactLinks = () => conf.get(REACT_LINKS_COUNTER);
exports.saveNumberOfReactLinks = (count) => conf.set(REACT_LINKS_COUNTER, count);
const envFromProcessEnv = {
    prod: Environment.Production,
};
exports.getEnvironment = () => {
    const env = envFromProcessEnv[process.env.VTEX_ENV];
    const persisted = conf.get('env') || exports.ENV_DEFAULT_VALUE;
    return env || persisted;
};
var Region;
(function (Region) {
    Region["Production"] = "aws-us-east-1";
})(Region = exports.Region || (exports.Region = {}));
exports.saveCluster = (cluster) => {
    conf.set('cluster', cluster);
};
exports.getCluster = () => {
    return conf.get('cluster') || exports.CLUSTER_DEFAULT_VALUE;
};
